thalassa/api.py

```diff
@@ -65,24 +65,102 @@ def get_elevation_dmap(trimesh: gv.TriMesh, show_grid: bool = False) -> hv.Overl
-    def get_timeseries_xy(x,y):
-        dist=abs(sxi+1j*syi-x-1j*y); nid=np.nonzero(dist==dist.min())[0][0]
-        mti=fdata['time'].data; myi=fdata['elev'].data[:,nid].copy()
-        ha=hv.Curve((mti,myi),'time','elevation')
-        return ha
-
-    #fixed station
-    xys=((3.0,37),(30,33)) #todo: read fixed stations from inputs
-    hs=hv.HoloMap({'P_{}'.format(i+1):get_timeseries_xy(*xy) for i,xy in enumerate(xys)})
-
-    #dynamic
-    hd=hv.DynamicMap(get_timeseries_xy,streams=[PointerXY(x=0,y=0,source=source)])
-    hd=hd.opts(color='k',line_width=2,line_dash='dotted')
-
-    #points
-    hp=hv.HoloMap({'P_{}'.format(i+1):hv.Points(xy).opts(color='r') for i,xy in enumerate(xys)})
-
-    return hd*hs.overlay(),hp.overlay()
+#----------------------------------------------------------------------------------------
+#time series
+#----------------------------------------------------------------------------------------
+#define a class to store data related to time series points
+class timeseries_data:
+      def __init__(self):
+         self.init=False
+      def clear(self):
+          self.init=False
+
+#function for extracting time series@(x,y) from data
+def extract_timeseries(x,y,sx,sy,data):
+    dist=abs(sx+1j*sy-x-1j*y)
+    mdist=dist.min()
+    nid=np.nonzero(dist==mdist)[0][0]
+    mdata=data['elev'].data[:,nid].copy()
+    return mdist,mdata
+
+#function to dynamically add or remove pts by double clicking on the map
+def add_remove_pts(x,y,data,dataset,fmt):
+    if fmt=='add pts':
+       if len(data.xys)==0:
+          mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+          hcurve=hv.Curve((data.time,mdata),'time','elevation')
+          if mdist<=data.mdist:
+             data.xys.append((x,y))
+             data.elev.append(mdata)
+             data.curve.append(hcurve)
+       else:
+          if data.xys[-1][0]!=x and data.xys[-1][1]!=y:
+             mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+             hcurve=hv.Curve((data.time,mdata),'time','elevation')
+             if mdist<=data.mdist:
+                data.xys.append((x,y))
+                data.elev.append(mdata)
+                data.curve.append(hcurve)
+    elif fmt=='remove pts':
+       if len(data.xys)>0:
+          xys=np.array(data.xys)
+          dist=abs(xys[:,0]+1j*xys[:,1]-x-1j*y)
+          mdist=dist.min()
+          if mdist<=data.mdist:
+             nid=np.nonzero(dist==mdist)[0][0]
+             data.xys=[k for i,k in enumerate(data.xys) if i!=nid]
+             data.elev=[k for i,k in enumerate(data.elev) if i!=nid]
+             data.curve=[k for i,k in enumerate(data.curve) if i!=nid]
+    else:
+       pass
+
+#get time series plots
+def get_timeseries(source,data,dataset,ymin,ymax,fmt):
+    #initialize timeseries_data
+    if data.init is False:
+       #find the maximum side length
+       x,y=dataset['SCHISM_hgrid_node_x'].data,dataset['SCHISM_hgrid_node_y'].data
```

Comment:
1. Please one assignment per line
2. `SCHISM_hgrid_node_x` is hardcoded. This means that the timeseries functionality can only be used with SCHISM output. Trying to use it with anything else will just throw an exception. Please, either pass the names of the variables as arguments to this function, or directly pass the arrays as arguments